### PR TITLE
New reference panel: handle `codeIntel.mixPreciseAndSearchBasedReferences`

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes one of the last big chunks of https://github.com/sourcegraph/sourcegraph/issues/30970 by replicating the behaviour in `code-intel-extensions` where [we mix precise and search-based results](https://github.com/sourcegraph/code-intel-extensions/blob/1b19eb6e6ef2435d9f0f7c7639e1749787799e4b/template/src/providers.ts#L309-L341) depending on the user setting.

It works (see video below), but I ran into two issues that I think can be fixed as follow-ups:

- https://github.com/sourcegraph/sourcegraph/issues/33412 - this is a bug that we should fix
- https://github.com/sourcegraph/sourcegraph/issues/33417 - this might be nice to have

Video:

https://user-images.githubusercontent.com/1185253/161749404-facc90c1-226a-4b39-97b9-3a7cad3a9424.mp4

## Test plan

Tested this locally (see video). With different pagination sizes and in repos with and without precise code intel.

## App preview:
- [Link](https://sg-web-mrn-mix-search-precise.onrender.com)

### CODE REVIEW

File: client/web/src/codeintel/ReferencesPanel.tsx [hunk 1]
```diff
@@ -684,12 +684,8 @@ const CollapsibleRepoLocationGroup: React.FunctionComponent<{
     getLineContent: (location: Location) => string
     filter: string | undefined
 }> = ({ repoLocationGroup, setActiveLocation, getLineContent, activeLocation, filter }) => {
-    const allSearchBased = useMemo(
-        () =>
-            find(
-                repoLocationGroup.referenceGroups.flatMap(reference => reference.locations),
-                location => !location.precise
-            ) !== undefined,
+    const quality = useMemo(
```
Comment (efritz) [inline client/web/src/codeintel/ReferencesPanel.tsx:687]: One quality per location _group_? What is the group here? Previously, we knew that each file will either have precise _or_ search but will not mix within the same file.

Comment (mrnugget) [reply client/web/src/codeintel/ReferencesPanel.tsx:687]: Good point! Right now I group per repo (see video above). But showing this per file might make more sense now.

**Comment by mrnugget** [comment]
Pagination is fixed (see comment here for details: https://github.com/sourcegraph/sourcegraph/issues/33417#issuecomment-1090034260). Short version: only once LSIF data is exhausted do we add search-based results (if setting is enabled).

I also moved the search-based/precise labels to the file level.

**Comment by mrnugget** [comment]
There's a tiny bug in here that I need to fix: duplicate results.

**Review by efritz — APPROVED** [review]

**Pull request merged**